### PR TITLE
Provide a way for custom derives to know if they were invoked via `#[derive_const]`

### LINKED ISSUE #118304: Provide a way for derives to know if they were invoked with `#[derive_const]`
As part of #67792, `#[derive_const]` is the current way to indicate that a derive should be `impl const Trait` rather than `impl Trait`. However, this only works for built-in macros. It should be possible for end-user proc macros to know if they were invoked with `#[derive_const]`, permitting the derive to emit the appropriate code.

In my opinion, this could be an opaque type provided as an additional parameter to the derive implementation. That type would have methods to obtain constness and any other future extension (i.e. effects).

cc @fee1-dead per [this Zulip comment](https://rust-lang.zulipchat.com/#narrow/stream/122651-general/topic/.22collaborator.22.20status.3F/near/404148155).

@rustbot label +A-proc-macros +C-enhancement +F-const-trait-impl +T-libs-api

### PR DESCRIPTION
**Pull request by fmease**
Fixes #118304.

cc @rust-lang/wg-macros
r? compiler and libs-api

### CODE REVIEW

File: tests/ui/rfcs/rfc-2632-const-trait-impl/const_derives/signature-proc-macro-derive.stderr [hunk 1]
```diff
@@ -0,0 +1,16 @@
+error: derive proc macro has incorrect signature
+  --> $DIR/signature-proc-macro-derive.rs:16:1
+   |
+LL | pub fn bad_input() -> proc_macro::TokenStream {
+   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+   | |
+   | incorrect number of function parameters
+   | incorrect number of function parameters
```
Comment (fmease) [inline tests/ui/rfcs/rfc-2632-const-trait-impl/const_derives/signature-proc-macro-derive.stderr:8]: duplication sad but annoying to address (would need to modify `note_type_err` to deal with this special usage)

**Comment by fmease** [comment]
Ah, I need to update r-a, too (`proc-macro-srv/src/server.rs`).

**Comment by fmease** [comment]
> rust-analyzer is developed in its own repository. If possible, consider making this change to [rust-lang/rust-analyzer](https://github.com/rust-lang/rust-analyzer) instead.

Pretty that this in-tree change is unavoidable.

**Comment by fmease** [comment]
@rust-lang/libs-api, would you like to see an ACP for this first?

**Comment by petrochenkov** [comment]
(I should get to this on new year holidays.)

File: compiler/rustc_passes/src/check_attr.rs [hunk 2]
```diff
@@ -2368,18 +2391,33 @@ impl<'tcx> CheckAttrVisitor<'tcx> {
...
+            };
+
+            note_expected_found(expected_sig);
+
+            if let ProcMacroKind::Derive = kind
+                && tcx
+                    .features()
+                    .declared_lib_features
+                    .iter()
+                    .any(|&(feature, _)| feature == sym::derive_const)
```
Comment (fmease) [inline compiler/rustc_passes/src/check_attr.rs:2416]: I think I should actually be able to do:

```suggestion
                && tcx.features().declared(sym::derive_const)
```

or is it

```suggestion
                && tcx.features().active(sym::derive_const)
```

File: library/proc_macro/src/bridge/client.rs [hunk 3]
```diff
@@ -506,3 +510,26 @@ impl ProcMacro {
...
+        self
+    }
+}
+
+impl<F> const ExpandCustomDerive<crate::DeriveExpansionOptions> for F
+where
+    F: Fn(crate::TokenStream, crate::DeriveExpansionOptions) -> crate::TokenStream + Copy,
+{
+    fn into_fn(self) -> impl Fn(crate::TokenStream) -> crate::TokenStream + Copy {
+        move |input| self(input, Default::default())
```
Comment (fmease) [inline library/proc_macro/src/bridge/client.rs:533]: This feels a bit weird and it demonstrates that `DeriveExpansionOptions` is a “sham”. I fear that the way it's wired right now, this impl isn't extensible. Consider the hypothetical in which we *do* store something meaningful in `DeriveExpansionOptions`, we wouldn't be able to retrieve it, well with `FreeFunctions`we could I guess. Maybe this is a non-issue, I'm just rambling at this point ^^'

**Comment by fmease** [comment]
Nominating for t-libs-api

> TL;DR: Allow derive-proc-macros to have the signature `fn(TokenStream, DeriveExpansionOptions) -> TokenStream` where `DeriveExpansionOptions` is an abstract type with the method `fn is_const(&self) -> bool` which returns whether the macro was invoked via `#[derive_const]` or `#[derive]`. Afaik, `#[derive_const]` isn't backed by an RFC just like the rest of `const_trait_impl`/`effects`. Just happy to receive feedback from you if you're roughly okay with the shape of this unstable API or not :)

**Comment by petrochenkov** [comment]
Both `#[derive_const]` itself, and this, are maximally ad hoc tools for addressing the issue.

How macros currently take their input - by a token stream, or two token streams if the macro has two inputs like `#[attr(input1)] INPUT2`.

This PR adds a third backdoor input tailored for a very specific use case.
The `#[derive_const]` added a whole new macro just to add one configuration bit to already existing `#[derive]` macro.

I would suggest keeping the current model and using already existing token-based methods for configuring macro behaviors.
- Either helper attributes, like many derive macros, including `derive_more`, already do
- Or giving derive macros the second input token stream - `#[derive(Clone(input1))] INPUT2`, that would also improve consistency with attributes.

Then the macros themselves can process the inputs as they want.
E.g. built-in macros could conventionally take a `const` token to produce a const impl - `#[derive(Clone(const))] struct S;`, others may accept more configuration.

If `#[derive]` needs some sugar to pass `const` to multiple derive macros that it applies at once, then it can grow such sugar itself, instead of delegating to a whole new special-purpose macro.
E.g. `#[derive(const: Clone, PartialEq, PartialOrd)] struct S;` or anything else, tokens inside the parentheses are an arbitrary token stream, I remind you, so `#[derive]` can create its own language for its input, like any other attribute macro.

**Comment by petrochenkov** [comment]
On a second thought, there's a third way to pass inputs - the "global data" in `struct Rustc`.

Proc macro methods without arguments like `Span::call_site()` take their return values from that global data.
I also planned to add some more data there to be available to proc macros, like the kind of brackets used by the macro.

If "derive expansion options" are placed into `struct Rustc`, then there's no need to add a second parameter to `#[proc_macro_derive]` functions for passing them.

**Comment by petrochenkov** [comment]
What other similar configuration values we may need in the future?

Will we need to pass them to derive macros only, or for attribute possibly for attribute or fn-like macros too?
With derive macros we have an intermediate layer, the `#[derive]` macro which has access to compiler magic to set such configuration values, but for attribute and fn-like macros we do not.

**Comment by Dylan-DPC** [comment]
@fmease any updates on this? thanks

**Comment by fmease** [comment]
No updates yet, my plan is to discuss petrochenkov's very good feedback with the rest of project-const-traits as well as with wg-macros.
@rustbot blocked

**Comment by fmease** [comment]
I've updated #118304 with petrochenkov's valid criticism and I will close this PR instead of keeping it open until a redesign of this feature to preserve history.
